src/portable/portable.c

```
@@ -465,23 +400,27 @@ static int portable_extract_by_path(
 
                 for (;;) {
                         _cleanup_(portable_metadata_unrefp) PortableMetadata *add = NULL;
-                        _cleanup_free_ char *name = NULL;
                         _cleanup_close_ int fd = -1;
+                        char iov_buffer[PATH_MAX + 2];
+                        struct iovec iov = IOVEC_INIT(iov_buffer, sizeof(iov_buffer));
 
-                        r = recv_item(seq[0], &name, &fd);
-                        if (r < 0)
-                                return log_debug_errno(r, "Failed to receive item: %m");
+                        ssize_t n = receive_one_fd_iov(seq[0], &iov, 1, 0, &fd);
+                        if (n == -EIO)
```

why the special handling of EIO?

---

It's the one difference with the previous local implementation, it returns eio when there's no more messages

---

hmm, not sure i follow? where precisely is this EIO supposed to come from?

---

It's this one here https://github.com/systemd/systemd/blob/fa4f366211ecbd123a067d568a8d3010024aaab8/src/basic/socket-util.c#L1013

---

hmm, we probably should return a better error there, given that EIO might come from any other result too. maybe change receive_one_fd_iov() to return ENOMSG in that case?

i did a quick check and it suggests no current caller (besides one test case) cares about the return value, so we should be able to just change it. but maybe double check

---

I'm not too confident it's safe to change, as it's bound to a varlink API: receive_one_fd_iov -> dynamic_user_pop -> dynamic_user_current -> vl_method_get_user_record -> io.systemd.UserDatabase.GetUserRecord

---

there should be a comment here about why EIO is matched at the very least